src/main/java/org/folio/circulation/domain/validation/ItemLimitValidator.java

```diff
@@ -0,0 +1,103 @@
+package org.folio.circulation.domain.validation;
+
+import static java.util.concurrent.CompletableFuture.completedFuture;
+import static org.folio.circulation.support.Result.ofAsync;
+import static org.folio.circulation.support.Result.succeeded;
+
+import java.util.concurrent.CompletableFuture;
+import java.util.function.Function;
+
+import org.apache.commons.lang3.StringUtils;
+import org.folio.circulation.domain.Item;
+import org.folio.circulation.domain.Loan;
+import org.folio.circulation.domain.LoanAndRelatedRecords;
+import org.folio.circulation.domain.LoanRepository;
+import org.folio.circulation.rules.AppliedRuleConditions;
+import org.folio.circulation.support.Result;
+import org.folio.circulation.support.ValidationErrorFailure;
+
+public class ItemLimitValidator {
+  private final Function<String, ValidationErrorFailure> itemLimitErrorFunction;
+  private final LoanRepository loanRepository;
+  private final static int LOANS_LIMIT = 10000;
+
+  public ItemLimitValidator(Function<String, ValidationErrorFailure> itemLimitErrorFunction,
+    LoanRepository loanRepository) {
+
+    this.itemLimitErrorFunction = itemLimitErrorFunction;
+    this.loanRepository = loanRepository;
+  }
+
+  public CompletableFuture<Result<LoanAndRelatedRecords>> refuseWhenItemLimitIsReached(
+    LoanAndRelatedRecords records) {
+
+    Loan loan = records.getLoan();
+    Integer itemLimit = loan.getLoanPolicy().getItemLimit();
+
+    if (itemLimit == null) {
+      return completedFuture(succeeded(records));
+    }
+
+    return ofAsync(() -> loan.getLoanPolicy().getRuleConditions())
+      .thenComposeAsync(result -> result.failAfter(ruleConditions -> isLimitReached(ruleConditions, records),
+        ruleConditions -> {
+          String message = getErrorMessage(ruleConditions);
+          return itemLimitErrorFunction.apply(String.format("Patron has reached maximum item limit of %d items %s",
+            itemLimit, message));
+        }))
+      .thenApply(result -> result.map(v -> records));
+  }
+
+  private CompletableFuture<Result<Boolean>> isLimitReached(
+    AppliedRuleConditions ruleConditionsEntity, LoanAndRelatedRecords records) {
+
+    if (!ruleConditionsEntity.isItemTypePresent() && !ruleConditionsEntity.isLoanTypePresent()) {
+      return ofAsync(() -> false);
+    }
+
+    Item item = records.getLoan().getItem();
+    String materialTypeId = item.getMaterialType() != null
+      ? item.getMaterialTypeId()
+      : null;
+    String loanTypeId = item.determineLoanTypeForItem();
```

Comment:
Why is this referring to the item, rather than the applicable rule?

My understanding is that the loans that are included in the count are based upon the applicable rule, rather than the item being checked out itself.

I understand that these values are likely to be the same, due to how the circulation rules are structured, making this comment more about reveal the intent of the logic, rather than whether it behaves correctly.

Reply:
The intent of the logic was to compare item materialType/loanType with retrieved loans to filter them. 
Rules was used only to indicate if it is present